### PR TITLE
use device-agnostic PyTorch accelerator APIs

### PR DESCRIPTION
**Pull request by Madhustat**
Relates to #1303
## Summary 
Replace CUDA-specific device-management calls with PyTorch's device-agnostic `torch.accelerator` APIs.

This removes CUDA assumptions from common OpenRLHF code paths and allows the affected code paths to use supported accelerators, including NVIDIA CUDA and Intel XPU, without vendor-specific branches.

This PR is intentionally limited to direct device-API substitutions. Feature-specific changes related to ring attention, and vLLM weight synchronization are excluded and will be submitted separately.

## Motivation

Several common OpenRLHF code paths use CUDA-specific APIs for operations that are not inherently CUDA-specific, including:

- Selecting the current device
- Retrieving the current device index
- Clearing unused accelerator memory
- Synchronizing accelerator operations
- Reading device properties
- Creating device meshes

These calls prevent the affected paths from running with a non-CUDA PyTorch build, even when the active accelerator provides equivalent functionality.

PyTorch's `torch.accelerator` namespace provides a common interface for device management across supported accelerator backends. This PR uses that interface where the underlying operation is equivalent across devices.

### Device-management operations

Apply the following direct substitutions:

```text
torch.cuda.empty_cache()
    -> torch.accelerator.empty_cache()

torch.cuda.synchronize()
    -> torch.accelerator.synchronize()

torch.cuda.set_device(local_rank)
    -> torch.accelerator.set_device_index(local_rank)
```

### Device-index operations

Replace:

```python
torch.cuda.current_device()
```

with:

```python
torch.accelerator.current_device_index()
```

Both APIs return the integer index of the currently selected device.

This intentionally does not use:

```python
torch.accelerator.current_accelerator()
```

`current_accelerator()` identifies the accelerator device type, whereas `current_device_index()` preserves the selected device index required for correct multi-device placement.

### Device-type operations

Use the active accelerator type when an API requires a device-type string:

```python
accelerator_type = torch.accelerator.current_accelerator().type

device_mesh = init_device_mesh(
    accelerator_type,
    mesh_shape,
)
```

### Device-property lookup

`torch.cuda.get_device_properties()` does not have a direct generic equivalent under `torch.accelerator`.

Resolve the active accelerator module and call its backend-specific property
API:

```python
accelerator = torch.accelerator.current_accelerator()
accelerator_module = getattr(torch, accelerator.type)

properties = accelerator_module.get_device_properties(
    torch.accelerator.current_device_index()
)
```

This preserves backend-specific device-property access without hard-coding `torch.cuda`.

## Files Changed

This PR modifies seven files:

```text
openrlhf/trainer/ray/launcher.py
openrlhf/trainer/ray/ppo_critic.py
openrlhf/trainer/ray/utils.py
openrlhf/trainer/ppo_utils/replay_buffer.py
openrlhf/utils/deepspeed/deepspeed.py
openrlhf/utils/deepspeed/deepspeed_utils.py
openrlhf/utils/loss_utils.py
```

Each file contains only device-API portability changes.

Device-API changes inside feature-specific files, including ring attention, vLLM device placement, and vLLM weight synchronization, are intentionally excluded and will be submitted with their respective changes.

### NVIDIA CUDA

Validated from a clean checkout of this PR on a multi-GPU NVIDIA system.

Validation environment:

```text
PyTorch: 2.11.0+cu130
```

The repository test suite passes:

```bash
python -m pytest tests
```

The following end-to-end workflows completed successfully:

- GRPO using `group_norm`
- PPO with a critic using `gae`

The tests exercised:

- Device selection and device-index handling
- Actor and critic forward passes
- Rollout generation
- Backward propagation
- Optimizer updates
- Cache cleanup
- Accelerator synchronization
- Device-mesh creation
- Device-property lookup

The migrated APIs preserved the behavior of the original CUDA-specific calls in the tested workflows. No regressions were observed.

### Intel XPU

Validated the migrated device-management APIs on Intel XPU.

Validation environment:

```text
PyTorch: 2.12.0+xpu
```

The following behavior was confirmed:

- `current_device_index()` returns the active backend device index
- `set_device_index(i)` binds execution to the selected XPU device
- `current_accelerator().type` resolves the device type used for device-mesh creation
- `get_device_properties()` resolves through the active accelerator module
- Device indices are preserved across the tested XPU devices

Full XPU end-to-end training requires additional feature-specific compatibility changes outside the scope of this PR, including the follow-up attention-import work.

End-to-end XPU results are therefore reported with the combined enablement changes rather than attributed to this device-API-only PR.

### Multi-device behavior

The device-index substitutions preserve the integer device index returned by the original CUDA APIs.

In particular:

```text
torch.cuda.current_device()
    -> torch.accelerator.current_device_index()
```

This preserves the distinction between device indices such as device 0 and device 1 rather than reducing the value to only the accelerator type.

Device-index preservation was validated on the tested multi-device CUDA and XPU configurations, with each rank bound to its assigned physical device index.

### NVIDIA CUDA

On a CUDA-enabled PyTorch build, `torch.accelerator` delegates the migrated operations to the active CUDA backend.

The tested NVIDIA workflows completed successfully without requiring CUDA-specific branches in the migrated code.

### Intel XPU

On an XPU-enabled PyTorch build, the same code resolves to the Intel XPU backend without requiring CUDA-specific calls or explicit vendor branching.

### CPU-only execution

This PR targets accelerator-backed execution paths.

The modified call sites are expected to run after an accelerator has been selected and initialized. This PR does not add a CPU fallback for accelerator-only training workflows.

## Scope

This PR changes device-API selection only.

It does not change:

- PPO, GRPO, or reward-model algorithms
- Loss calculations
- Model architecture
- Dataset processing
- Ray scheduling
- DeepSpeed configuration
- vLLM engine behavior
- Distributed communication backends

No new dependency is introduced.

### NVIDIA CUDA

- [x] Repository test suite passes on PyTorch `2.11.0+cu130`
- [x] GRPO using `group_norm` completes end to end
- [x] PPO with a critic using `gae` completes end to end
- [x] Real forward, backward, and optimizer-update execution validated
- [x] Multi-GPU device-index preservation confirmed
- [x] No regressions observed in the tested CUDA workflows

### Intel XPU

- [x] Migrated device APIs validated on PyTorch `2.12.0+xpu`
- [x] Device-type resolution confirmed
- [x] Device-property lookup confirmed
- [x] Device-index preservation confirmed across the tested XPU devices
- [x] Multi-device rank-to-device binding confirmed

### Code Quality

- [x] `pre-commit run --all-files` passes

### CODE REVIEW

**Comment by Madhustat** [comment]
Hi @hijkzzz, @Excelius-Wang

This is the core device-portability change for running OpenRLHF workflows across CUDA and XPU without maintaining separate code paths.
The affected upstream workflows are passing on both single and multi XPU. 
If there’s anything I can do to help with the review and merge of this PR, I’m more than happy to work on it.

Regards,
Madhu
